runtime/src/transact_asset.rs

```diff
@@ -0,0 +1,120 @@
+#![allow(clippy::all)]
+use codec::FullCodec;
+use sp_runtime::traits::{Convert, MaybeSerializeDeserialize};
+use sp_std::{
+    cmp::{Eq, PartialEq},
+    fmt::Debug,
+    marker::PhantomData,
+    prelude::*,
+    result,
+};
+use xcm::v0::{Error as XcmError, MultiAsset, MultiLocation, Result};
+use xcm_executor::{
+    traits::{Convert as MoreConvert, MatchesFungible, TransactAsset},
+    Assets,
+};
+
+// Asset transaction errors.
+enum Error {
+    /// Failed to match fungible.
+    FailedToMatchFungible,
+    /// `MultiLocation` to `AccountId` Conversion failed.
+    AccountIdConversionFailed,
+    /// `CurrencyId` conversion failed.
+    CurrencyIdConversionFailed,
+}
+
+impl From<Error> for XcmError {
+    fn from(e: Error) -> Self {
+        match e {
+            Error::FailedToMatchFungible => {
+                XcmError::FailedToTransactAsset("FailedToMatchFungible")
+            }
+            Error::AccountIdConversionFailed => {
+                XcmError::FailedToTransactAsset("AccountIdConversionFailed")
+            }
+            Error::CurrencyIdConversionFailed => {
+                XcmError::FailedToTransactAsset("CurrencyIdConversionFailed")
+            }
+        }
+    }
+}
+
+/// The `TransactAsset` implementation, to handle `MultiAsset` deposit/withdraw.
+/// Note that teleport related functions are unimplemented.
+///
+/// If the asset is known, deposit/withdraw will be handled by `MultiCurrency`,
+/// else by `UnknownAsset` if unknown.
+pub struct MultiCurrencyAdapter<
```

Comment:
Don't love the way this is sooo generic having to deal with all the type gymnastics but not sure if it's avoidable.
Also since we don't use `MultiCurrency` in our system perhaps a better name could be something like `AssetsAdapter`? or `NetworkAssetsAdapter`?